crates/dbsp/src/operator/trace.rs

```diff
@@ -230,6 +276,145 @@ where
+    /// consumers.
+    ///
+    /// The second mechanism, implemented by this method and the
+    /// [`integrate_trace_retain_values`](`Self::integrate_trace_retain_values`)
+    /// method, is more general and allows the caller to specify an
+    /// arbitrary condition on keys and values in the trace respectively.
+    /// Keys or values that don't satisfy the condition are eventually
+    /// reclaimed by the trace.  This mechanism is applicable to collections
+    /// that are not ordered by time.  Hence it doesn't require rearranging
+    /// the data in time order.  Furthermore, it is applicable to collections
+    /// that contain multiple timestamp column.  Such multidimensional
+    /// timestamps only form a partial order.
+    ///
+    /// Unlike the first mechanism, this mechanism only allows one global
+    /// condition to be applied to the stream.  This bound affects _all_
+    /// operators that use the trace of the stream, i.e., call
+    /// `integrate_trace` (or `trace` in the root scope) on it. This includes
+    /// for instance `join`, `aggregate`, and `distinct`.  All such operators
+    /// will reference the same instance of a trace.  Therefore bounds
+    /// specified by this API must be based on a global analysis of the
+    /// entire program.
+    ///
+    /// The two mechanisms described above interact in different ways for keys
+    /// and values. For keys, the lower bound and the retainment condition
+    /// are independent and can be active at the same time.  Internally,
+    /// they are enforced using different techniques. Lower bounds are
+    /// enforced at essentially zero cost.  The retention condition is more
+    /// expensive, but more general.
+    ///
+    /// For values, only one of the two mechanisms can be enabled for any given
+    /// stream.  Whenever a retainment condition is specified it supersedes
+    /// any lower bounds constraints.
+    ///
+    /// ## Arguments
+    ///
+    /// * `watermark_stream` - This stream carries scalar values (i.e., single
+    ///   records, not Z-sets).  The key retainment condition is defined
+    ///   relative to the last value received from this stream. Typically, this
+    ///   value represents the lowest upper bound of all partially ordered
+    ///   timestamps in `self` or some other stream, computed with the help of
+    ///   the [`watermark`](`Stream::watermark`) operator and adjusted by some
+    ///   contstant offsets, dictated, e.g., by window sizes used in the queries
+    ///   and the maximal out-of-ordedness of data in the input streams.
+    ///
+    /// * `retain_key_func` - given the value received from the
+    ///   `watermark_stream` at the last clock cycle and a key, returns `true`
+    ///   if the key should be retained in the trace and `false` if it should be
+    ///   discarded.
+    ///
+    /// ## Correctness
+    ///
+    /// * As discussed above, the retainment policy set using this method
```

Comment:
moreover, this requires certain assumptions about the inputs, i.e., lateness.
Can we rename the watermark?

Reply:
No objections, but I don't have a better name in mind.  And "watermark" kind of makes sense. What would you call it?

Reply:
Lateness?
Let's not muddle this further

Reply:
This really isn't the same as lateness.  But just to clarify, are you suggesting renaming the `watermark` operator or the `watermark_stream` argument to this function.  In practice, the value in the `watermark_stream` will be computed in such a way that `retain_func` will simply be `|ts, key| key>=ts`, so that the amount of work per key is minimal.  So this is really more like a bound than lateness.